### PR TITLE
Rebalances and adds more ambush-style cling abilities, buffs chameleon skin and adds a new ability! "Darkness Adaptation"

### PR DESCRIPTION
**Pull request by Singul0**
## About The Pull Request
This PR adds a new ambush ability to changelings, "Darkness Adaptation" and tweaks the balance of "Chameleon Skin".

The new cling ability, "Darkness Adaptation".  makes your character translucent and slightly dark while also giving the ability to see slightly better into dark than normally (30% better). While this ability is active, flash protection is lowered by one.

Buffs "Chameleon Skin" by reducing the DNA cost to 1 from 2, and sped up the time it takes to go full invisible by 2.5x and zero instability cost

Below is the picture of the translucency through normal non-night-vision vision: (note: this is using extremely dark armor already)
gambar

Below is the picture of the translucency through the poor night-vision that the ability gives you:
gambar
## Why It's Good For The Game
This ability was added in #11148 back in 2015 and hasn't been touched for 9 years. The original ability just gave you the _extremely_ situational "chameleon skin" mutation (one that is accessible to the crew through genetics already). for a whopping **2 mutation points** cost. The mutation makes you invisible after half a minute or so of standing still and is immediately broken the second you move or get pushed.

Compared to other similarly costed powers you get for 2 mutation points, this is by far the worst bang for your buck by a large margin.

Needless to say, considering with how power-crept everything else in the game is after all these years. this ability that hasn't been touched for ages have fell into disuse to the point of uselessness, especially with the heavy cost it that came with it.

This PR buffs that ability by making it cheaper, in line with the cost of other less powerful abilities that clings have. and made it 2.5x faster to turn invisible.

The PR also adds another ambush-oriented ability, "Darkness adaptation". The ability allows your character to become translucent and slightly dark while also giving the ability to see slightly better into dark than normally (30% better); with the downside of your eyes being slightly worse than before.

This ability follows the general spirit of clings, an stealth-generalist oriented antag that you never know when they would strike (disguises already have this effect, but them being far more able to jump you in maint even more so.). and encourages more creative strategy than just your average murderbone. 

A Perfect force multiplier for maint combat and maint ambushes. Also synergises well with augmented sight, and the blackish cling armor.

Don't worry if you think the new ability, combined with the faster cham skin would make you invisible to the naked eye. The mutation cancels out the translucency of the adaptation.
## Changelog
:cl:
add: Adds a new changeling ability, "Darkness Adaptation". Making you more translucent, especially in darkness and allowing you to see slightly better in the dark
balance: The changeling power "Chameleon Skin" has been buffed, Reduces the cost to 1 and sped up the time it takes to turn invisible
/:cl:

### CODE REVIEW

**Comment by DaCoolBoss** [comment]
Night vision seems redundant with augmented eyesight also available.

**Comment by Singul0** [comment]
> Night vision seems redundant with augmented eyesight also available.

Augmented eyesight only gives you xray.

File: code/modules/antagonists/changeling/powers/chameleon_skin.dm [hunk 1]
```diff
@@ -1,25 +1,43 @@
...
-	if(H.dna.get_mutation(/datum/mutation/human/chameleon))
-		H.dna.remove_mutation(/datum/mutation/human/chameleon)
+	if(is_active == FALSE)
+		animate(cling, alpha = 85,time = 3 SECONDS)
+		cling.visible_message("<span class='warning'>[cling.name] skin is suddenly slowly becoming translucent!</span>", \
+						"<span class='notice'>You are now far more stealthy and better at looking in the dark.</span>")
+		animate(cling, color = COLOR_GRAY, time = 3 SECONDS) //darkens their overall appearance
+		var/obj/item/organ/internal/eyes/eyes = cling.get_organ_by_type(/obj/item/organ/internal/eyes)
+		eyes.lighting_cutoff = LIGHTING_CUTOFF_MEDIUM // Adds barely usable, kinda shit night vision
+		eyes.flash_protect = max(eyes.flash_protect - 1, FLASH_PROTECTION_HYPER_SENSITIVE) //Reduces flash protection by one level.
```
Comment (MrMelbert) [inline code/modules/antagonists/changeling/powers/chameleon_skin.dm:23]: This has a funky order-of-operations bug with Augmented Eyesight that needs to be fixed. 

If activate this ability, and then de-activate augmented eyesight, the flash protection it offers will completely override this.

Rather than having `FLASH_PROTECTION_WELDER - 1` protection as you would expect, it would just have `FLASH_PROTECTION_WELDER`.  

Likewise, If you had this ability active and then activated augmented eyesight, you would have `FLASH_PROTECTION_SENSITIVE` eyes rather than `FLASH_PROTECTION_HYPER_SENSITIVE`.

File: code/modules/antagonists/changeling/powers/chameleon_skin.dm [hunk 2]
```diff
@@ -1,25 +1,43 @@
...
-		var/mob/living/carbon/C = user
-		C.dna.remove_mutation(/datum/mutation/human/chameleon)
 	..()
+	var/mob/living/carbon/human/cling = user
+	disable_ability(cling)
+
+/datum/action/changeling/chameleon_skin/proc/disable_ability(mob/living/carbon/human/cling) //Restores adaptation
+	animate(cling, alpha = 255, time = 3 SECONDS)
+	cling.visible_message("<span class='warning'>[cling.name] appears from thin air!</span>", \
+					"<span class='notice'>You are now appearing normal and lost the ability at looking in the dark.</span>")
```
Comment (MrMelbert) [inline code/modules/antagonists/changeling/powers/chameleon_skin.dm:38]: Few things: 
- Span macros
- Changings use "We" and not "You" 
- You don't need `.name`
- Minor grammar check 
```suggestion
	cling.visible_message(
		span_warning("[cling] appears from thin air!)",
		span_notice("We are now appearing normal and lost the ability to see in the dark."),
	)
```

File: code/modules/antagonists/changeling/powers/chameleon_skin.dm [hunk 3]
```diff
@@ -1,25 +1,43 @@
...
+	animate(cling, color = null, time = 3 SECONDS)
+	var/obj/item/organ/internal/eyes/eyes = cling.get_organ_by_type(/obj/item/organ/internal/eyes)
+	eyes.lighting_cutoff = LIGHTING_CUTOFF_VISIBLE
+	eyes.flash_protect = max(eyes.flash_protect + 1, FLASH_PROTECTION_WELDER)
```
Comment (MrMelbert) [inline code/modules/antagonists/changeling/powers/chameleon_skin.dm:42]: Also needs a sanity check that eyes exist.

File: code/modules/antagonists/changeling/powers/chameleon_skin.dm [hunk 4]
```diff
@@ -1,25 +1,43 @@
...
-	if(!istype(H)) // req_human could be done in can_sting stuff.
-		return
+	var/mob/living/carbon/human/cling = user //SHOULD always be human, because req_human = TRUE
 	..()
-	if(H.dna.get_mutation(/datum/mutation/human/chameleon))
-		H.dna.remove_mutation(/datum/mutation/human/chameleon)
+	if(is_active == FALSE)
+		animate(cling, alpha = 85,time = 3 SECONDS)
+		cling.visible_message("<span class='warning'>[cling.name] skin is suddenly slowly becoming translucent!</span>", \
+						"<span class='notice'>You are now far more stealthy and better at looking in the dark.</span>")
```
Comment (MrMelbert) [inline code/modules/antagonists/changeling/powers/chameleon_skin.dm:19]:
```suggestion
		cling.visible_message("
			span_warning("[cling]'s skin suddenly starts becoming translucent!",
			span_notice("We are now far more stealthy and better at seeing in the dark.",
		)
```

File: code/modules/antagonists/changeling/powers/chameleon_skin.dm [hunk 5]
```diff
@@ -1,25 +1,43 @@
...
+	name = "Darkness Adaptation"
+	desc = "Our skin pigmentation and eyes rapidly changes to suit the darkness. Needs 25 chemicals in-hand to toggle."
+	helptext = "Allows us to darken and change the translucency of our pigmentation, and adapt our eyes to see in dark conditions, \
+	The translucent effect works best in dark enviroments. Can be toggled on and off."
 	button_icon_state = "chameleon_skin"
 	dna_cost = 2
 	chemical_cost = 25
 	req_human = TRUE
+	/////is ability active?
+	var/is_active = FALSE
```
Comment (MrMelbert) [inline code/modules/antagonists/changeling/powers/chameleon_skin.dm:11]:
```suggestion
	/// is ability active (we are invisible)?
	var/is_active = FALSE
```

File: code/modules/antagonists/changeling/powers/chameleon_skin.dm [hunk 6]
```diff
@@ -1,25 +1,43 @@
...
 
 /datum/action/changeling/chameleon_skin/sting_action(mob/user)
-	var/mob/living/carbon/human/H = user //SHOULD always be human, because req_human = TRUE
-	if(!istype(H)) // req_human could be done in can_sting stuff.
-		return
+	var/mob/living/carbon/human/cling = user //SHOULD always be human, because req_human = TRUE
```
Comment (MrMelbert) [inline code/modules/antagonists/changeling/powers/chameleon_skin.dm:14]: FTR you can skip the middleman and just cast `user` to `human` in the parameter

**Review by MrMelbert — COMMENTED** [review]
Maybe remove the base chemical cost and replace it with a penalty to chemical cost regen like other similar passive-like abilities

Or a combination of both? Cheaper base cost, but a penalty to chem regen.

**Comment by Jezzared** [comment]
I actually really like Cham skin as an ambush tool. Popping cham skin and then leaning against a wall makes you invisible and passthrough and it was great for lurking in a spot waiting for a surprise attack.

This feels like its just worse than the original. I don't really know what use for this i'd have other than the night vision; you are SUPER visible in the current implementation of this.

**Comment by Jezzared** [comment]
> the idea of this, it seems, is that you are ambushing a random target, not security moth with 4 seclights

Let me rephrase this, then.

Who in their right mind isn't walking through maints without a flashlight? You get one on your PDA. It's shit, but it'll still reveal someone with this ability when it passes over them. The translucency is much lower than most other forms of this. You can be seen very well, even in low-light.

And if they aren't, who is going to see you in the darkness _anyways_? If they aren't using a flashlight they're open to you just jumping out of fulldark. And at that point all you're buying is the nightvision.

**Comment by necromanceranne** [comment]
Honestly, I just think the chameleon invisibility needs to just be faster rather than a total replacement. It'd be useful if only it was quicker.

File: code/modules/antagonists/changeling/powers/chameleon_skin.dm (hunk 1 above)
Comment (Singul0) [reply code/modules/antagonists/changeling/powers/chameleon_skin.dm:23]: I'll just change how augmented eyesight changes flash_protection to be in line withe how this new power works. is that okay?

Comment (Singul0) [reply code/modules/antagonists/changeling/powers/chameleon_skin.dm:25]: nevermind that just exacerbate the problem

Comment (MrMelbert) [reply code/modules/antagonists/changeling/powers/chameleon_skin.dm:25]: Probably fine, though a nerf to moth lings, which is based I guess

Comment (MrMelbert) [reply code/modules/antagonists/changeling/powers/chameleon_skin.dm:25]: Also you need to register signals for organ gained and organ lost so the eye resets its flash protection when eyes are removed and lost, like augmented eyesight.

Comment (Singul0) [reply code/modules/antagonists/changeling/powers/chameleon_skin.dm]: > Probably fine, though a nerf to moth lings, which is based I guess

sounds like a skill issue tbh

**Comment by Singul0** [comment]
Made the skin much darker and slightly more translucent than the ones in the picture.

**Comment by radarseas** [comment]
I'd say it would need more to it than just being slightly invisible in the dark but that's the nightmare's niché if anything this should be more of a nightmare exclusive ability than something for changlings.


Funnier idea... have it a sort of, disguise as a solid object as long as you stand still, a la morph? it would be funny as fuck to walk past a "chem dispenser" and suddenly get wallshoved and ganked by a ling that was disgusing as it, or to blend in as a "locker" in a security chase, then walk the other way as someone else out of sight.

**Comment by Singul0** [comment]
25% is a bit tad much yeah, i'll lower it to 15%

**Comment by bigfatbananacyclops** [comment]
should be added as an alternative option to chameleon skin instead of reworking it.

**Comment by Singul0** [comment]
> an alternative option to chameleon skin instead of reworking it.

as I said, I really can't. the codersprites the clings used to have has been replaced with new ones that I simply do not have the will nor expertise to match the quality of if I were to make this it's own ability.

**Comment by ExcessiveUseOfCobblestone** [comment]
Darkness ability seems cool if we didnt have a similar monster dedicated to the dark, Id rather the original power just be buffed and/or rebalanced to make it worth picking up more.

**Comment by Singul0** [comment]
> Darkness ability seems cool if we didnt have a similar monster dedicated to the dark

In my opinion, Overlapping abilities between isn't all that bad. This ability doesn't completely take over nightmare niche. It just allows clings to better perform in low-light enviroments such as maints and lavaland. Nowhere near the effectiveness of nightmares but considering cling's role as a generalist stealth antag similar to that of traitors it's warranted and justified for it to have some abilities that allow it to better perform in ambushes (low light enviroments are perfect for sneaky ambushes).

**Comment by Singul0** [comment]
Due to backlash and recommendations from here and discord, I have split the rework into it's own thing, and for the original ability to have a reduced cost and a faster fading time.

File: code/datums/mutations/chameleon.dm [hunk 7]
```diff
@@ -19,6 +19,11 @@
 /datum/mutation/human/chameleon/on_life(seconds_per_tick, times_fired)
 	owner.alpha = max(owner.alpha - (12.5 * (GET_MUTATION_POWER(src)) * seconds_per_tick), 0)
 
+//Upgraded mutation of the base variant, used for changelings. No instability and better power_coeff
+/datum/mutation/human/chameleon/changeling
+	instability = 0
+	power_coeff = 2.5
```
Comment (Ghommie) [inline code/datums/mutations/chameleon.dm:25]:
```suggestion
	power_coeff = 2.5
	locked = TRUE
```
We don't want this to be accessible by geneticists, right?

**Review by Ghommie — CHANGES_REQUESTED** [review]

**Comment by Singul0** [comment]
Checks is failing somehow? I'll push a commit later so that the check runs again

File: code/modules/antagonists/changeling/powers/darkness_adaptation.dm [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+	RegisterSignal(user, COMSIG_CARBON_LOSE_ORGAN, PROC_REF(eye_removed))
+
+/datum/action/changeling/darkness_adaptation/sting_action(mob/living/carbon/human/cling) //SHOULD always be human, because req_human = TRUE
+	..()
+	if(is_active)
+		disable_ability(cling)
+		is_active = FALSE
+	else
+		enable_ability(cling)
+		is_active = TRUE
```
Comment (Ghommie) [inline code/modules/antagonists/changeling/powers/darkness_adaptation.dm:28]:
```suggestion
	is_active = !is_active
	if(is_active)
		enable_ability(cling)
	else
		disable_ability(cling)
```
